<a href="https://colab.research.google.com/github/engbJapan/Cobla/blob/master/%E3%83%AD%E3%83%BC%E3%82%AB%E3%83%ABjupyter%E3%82%92%E6%8B%A1%E5%BC%B5%E3%81%A7%E4%B8%AD%E3%81%A7%E8%B5%B7%E5%8B%95%E3%81%AE%E4%BA%8B%EF%BC%92%EF%BC%92%EF%BC%91%EF%BC%91%EF%BC%91%EF%BC%98_%E3%81%93%E3%82%8C%E3%81%AF%E3%82%84%E3%82%8A%E6%90%8D.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# suテスト

In [1]:
!useradd -m engbjapan
!usermod -G root engbjapan
!id engbjapan

uid=1000(engbjapan) gid=1000(engbjapan) groups=1000(engbjapan),0(root)


In [2]:
!su -c 'id;ls -la;' engbjapan

uid=1000(engbjapan) gid=1000(engbjapan) groups=1000(engbjapan),0(root)
total 16
drwxr-xr-x 1 root root 4096 Nov 17 14:32 .
drwxr-xr-x 1 root root 4096 Nov 19 02:35 ..
drwxr-xr-x 4 root root 4096 Nov 17 14:31 .config
drwxr-xr-x 1 root root 4096 Nov 17 14:32 sample_data


In [3]:
!su - engbjapan -c "jupyter notebook --generate-config"

Writing default config to: /home/engbjapan/.jupyter/jupyter_notebook_config.py


In [28]:
!pip install jupyter_http_over_ws -qq
!su - engbjapan -c 'jupyter serverextension enable --py jupyter_http_over_ws'

     |████████████████████████████████| 1.6 MB 8.2 MB/s 
Enabling: jupyter_http_over_ws
- Writing config: /home/engbjapan/.jupyter
    - Validating...
      jupyter_http_over_ws 0.0.7 OK


In [ ]:
!cat /home/engbjapan/.jupyter/jupyter_notebook_config.py

In [6]:
%%bash
cat <<EPF >> /home/engbjapan/.jupyter/jupyter_notebook_config.py
"""Jupyter notebook server configuration for Colab runtimes."""

import logging


c.JupyterApp.answer_yes = True
c.NotebookApp.open_browser = False

c.NotebookApp.allow_root = True

c.Application.log_level = logging.DEBUG
c.NotebookApp.log_format = '|%(levelname)s|%(message)s'

c.NotebookApp.port_retries = 0

c.NotebookApp.login_handler_class = (
    'google.colab._login_handler.ColabLoginHandler')

c.MultiKernelManager.default_kernel_name = 'python3'

c.NotebookApp.allow_origin = '*'
c.NotebookApp.allow_remote_access = True
c.NotebookApp.disable_check_xsrf = True
c.NotebookApp.token = 'engbjapan'
c.Session.key = b''
c.Session.keyfile = ''
EPF

In [29]:
!cat /home/engbjapan/.jupyter/jupyter_notebook_config.py|egrep '^([^\\s|\#].*)'

"""Jupyter notebook server configuration for Colab runtimes."""
import logging
c.JupyterApp.answer_yes = True
c.NotebookApp.open_browser = False
c.NotebookApp.allow_root = True
c.Application.log_level = logging.DEBUG
c.NotebookApp.log_format = '|%(levelname)s|%(message)s'
c.NotebookApp.port_retries = 0
c.NotebookApp.login_handler_class = (
    'google.colab._login_handler.ColabLoginHandler')
c.MultiKernelManager.default_kernel_name = 'python3'
c.NotebookApp.allow_origin = '*'
c.NotebookApp.allow_remote_access = True
c.NotebookApp.disable_check_xsrf = True
c.NotebookApp.token = 'engbjapan'
c.Session.key = b''
c.Session.keyfile = ''


デフォルトのConfigは以下

In [10]:
import subprocess,shlex

def 実行(cmdlines:str,out=subprocess.PIPE):
  sb=subprocess.Popen(
    shlex.split(cmdlines.replace('\n',' '))
    ,stdout=out
    ,stderr=subprocess.STDOUT
  )
  return sb

#sb=実行("bash -c 'sleep 30;printf \"id\"' &")

#print(f"sb.pid:{sb.pid}")
#!ps -Af
#sb.communicate()[0]


In [34]:
sb=実行("su - engbjapan -c 'jupyter notebook --ip=127.0.0.1 --port=18888 --NotebookApp.allow_origin=\'https://colab.research.google.com\''")

pid=print(f"sb.pid:{sb.pid}")

#sb.stdout

#sb.kill()#効かねー
#!kill -9 1701
#!ps -Afl
"""
4 S root         233      75  0  80   0 - 20192 wait   02:44 ?        00:00:00 su - engbjapan -c jupyter notebook --ip=127.0.0.1 --port=18888
4 S engbjap+     234     233  0  80   0 -  1160 wait   02:44 ?        00:00:00 -su -c jupyter notebook --ip=127.0.0.1 --port=18888
4 S engbjap+     237
"""
#!kill -9 233
#!kill -9 234
#!kill -9 237

F S UID          PID    PPID  C PRI  NI ADDR SZ WCHAN  STIME TTY          TIME CMD
4 S root           1       0  0  80   0 -   252 sigtim 02:35 ?        00:00:00 /sbin/docker-init -- /datalab/run.sh
4 S root           7       1  0  80   0 - 86171 epoll_ 02:35 ?        00:00:04 /tools/node/bin/node /datalab/web/app.js
4 S root          18       0  0  80   0 -  8975 se_sys 02:35 ?        00:00:00 tail -n +0 -F /root/.config/Google/DriveFS/Logs/dpb.txt /root/.config/Google/Driv
1 S root          31       1  0  80   0 -  9802 wait   02:35 ?        00:00:00 /bin/bash -e /datalab/run.sh
4 S root          32      31  0  80   0 - 179957 futex_ 02:35 ?       00:00:00 /usr/colab/bin/kernel_manager_proxy --listen_port=6000 --target_port=9000
4 Z root          44       7  0  80   0 -     0 -      02:35 ?        00:00:06 [python3] <defunct>
4 S root          45       7  0  80   0 - 40377 epoll_ 02:35 ?        00:00:00 python3 /usr/local/bin/colab-fileshim.py
4 S root          60       7  0  80   0 

# うん？一応　notebook list出たね

In [33]:
!su - engbjapan -c 'jupyter notebook list'

Currently running servers:
http://127.0.0.1:18888/?token=engbjapan :: /home/engbjapan


## このtoken付URLをLocalランタイムのURLに入れても「接続」ボタンが出なかった

## 改行

In [ ]:
!su engbjapan -c \
  "cd /home/engbjapan/;\
  id;\
  pwd;\
  ls -la"
#!ls -la /home/engbjapan/

uid=1000(engbjapan) gid=1000(engbjapan) groups=1000(engbjapan),0(root)
/home/engbjapan
total 20
drwxr-xr-x 2 engbjapan engbjapan 4096 Nov 15 20:20 .
drwxr-xr-x 1 root      root      4096 Nov 15 20:20 ..
-rw-r--r-- 1 engbjapan engbjapan  220 Apr  4  2018 .bash_logout
-rw-r--r-- 1 engbjapan engbjapan 3771 Apr  4  2018 .bashrc
-rw-r--r-- 1 engbjapan engbjapan  807 Apr  4  2018 .profile


# エクステンションしてみる

```
以下のサイトだよ
https://research.google.com/colaboratory/local-runtimes.html

そのProxy拡張もオーブンで上のリンクに書いてある。
覗けば、webSocket系のリレーションproxyできる！！

Colab内部:
 上記extensionをpip：pip install jupyter_http_over_ws
 extension有効化ユーザで：
 su - {ユーザ} -c ”jupyter serverextension enable --py jupyter_http_over_ws”
 ユーザでjupyter起動:
 su - {ユーザ} -c \
  “jupyter notebook 
  --NotebookApp.allow_origin='https://colab.research.google.com' \
  --port=8888 \ ←これはportPickerので行くか
  --NotebookApp.port_retries=0”

　⇒これで出るアクセスURLを確認

Colabフロント：
　ローカルランタイム接続で、アクセスURL、を入れてランタイム変更
```

## 元メモ

https://docs.google.com/document/d/1QGFx4iXs6lhaZYiHTkyE0HYSLkqPj-fZpRLOONoImYE/edit?usp=sharing


In [ ]:
!pip install jupyter_http_over_ws

In [63]:
!apt install net-tools -qq

The following package was automatically installed and is no longer required:
  libnvidia-common-460
Use 'apt autoremove' to remove it.
The following NEW packages will be installed:
  net-tools
0 upgraded, 1 newly installed, 0 to remove and 5 not upgraded.
Need to get 194 kB of archives.
After this operation, 803 kB of additional disk space will be used.
Selecting previously unselected package net-tools.
(Reading database ... 123991 files and directories currently installed.)
Preparing to unpack .../net-tools_1.60+git20161116.90da8a0-1ubuntu1_amd64.deb ...
Unpacking net-tools (1.60+git20161116.90da8a0-1ubuntu1) ...
Setting up net-tools (1.60+git20161116.90da8a0-1ubuntu1) ...
Processing triggers for man-db (2.8.3-2ubuntu0.1) ...


In [32]:
!netstat -nap|grep LISTEN

/bin/bash: netstat: command not found


In [ ]:
!kill -9 1236

In [ ]:
!su - engbjapan -c "jupyter  serverextension enable --py jupyter_http_over_ws"

Enabling: jupyter_http_over_ws
- Writing config: /home/engbjapan/.jupyter
    - Validating...
      jupyter_http_over_ws 0.0.7 OK


In [ ]:
import portpicker
portpicker.pick_unused_port()

23415

## なんじゃこりゃ！！

```
バックエンドから .last_opt_in_prompt.yaml を取得できませんでした
Could not fetch resource at https://colab.research.google.com/tun/m/m-s-1kdrpsd5vooiu/files/content/.config/.last_opt_in_prompt.yaml?authuser=0: 500  
FetchError: Could not fetch resource at https://colab.research.google.com/tun/m/m-s-1kdrpsd5vooiu/files/content/.config/.last_opt_in_prompt.yaml?authuser=0: 500  
    at dC.fB [as constructor] (https://colab.research.google.com/v2/external/external_polymer_binary_l10n__ja.js?vrz=colab-20221117-060045-RC00_489179157:1405:2551)
    at new dC (https://colab.research.google.com/v2/external/external_polymer_binary_l10n__ja.js?vrz=colab-20221117-060045-RC00_489179157:1437:451)
    at wa.program_ (https://colab.research.google.com/v2/external/external_polymer_binary_l10n__ja.js?vrz=colab-20221117-060045-RC00_489179157:6458:358)
    at ya (https://colab.research.google.com/v2/external/external_polymer_binary_l10n__ja.js?vrz=colab-20221117-060045-RC00_489179157:20:336)
    at wa.next_ (https://colab.research.google.com/v2/external/external_polymer_binary_l10n__ja.js?vrz=colab-20221117-060045-RC00_489179157:18:479)
    at za.next (https://colab.research.google.com/v2/external/external_polymer_binary_l10n__ja.js?vrz=colab-20221117-060045-RC00_489179157:21:206)
    at b (https://colab.research.google.com/v2/external/external_polymer_binary_l10n__ja.js?vrz=colab-20221117-060045-RC00_489179157:21:468)
```

ならこれで行けんのか？

```
https://colab.research.google.com/tun/m/m-s-1kdrpsd5vooiu/files/content/ps.txt?authuser=0
```

# --ip 指定で動作したが、token情報が出ない


```
|WARNING|Notebook version 5 is no longer maintained. Please upgrade to version 6 or later.
***Debug***
        sockaddr('127.0.0.1', 8888)
|DEBUG|Paths used for configuration of jupyter_notebook_config: 
    	/etc/jupyter/jupyter_notebook_config.json
|DEBUG|Paths used for configuration of jupyter_notebook_config: 
    	/usr/local/etc/jupyter/jupyter_notebook_config.d/panel-client-jupyter.json
    	/usr/local/etc/jupyter/jupyter_notebook_config.json
|DEBUG|Paths used for configuration of jupyter_notebook_config: 
    	/usr/etc/jupyter/jupyter_notebook_config.json
|DEBUG|Paths used for configuration of jupyter_notebook_config: 
    	/home/engbjapan/.local/etc/jupyter/jupyter_notebook_config.json
|DEBUG|Paths used for configuration of jupyter_notebook_config: 
    	/home/engbjapan/.jupyter/jupyter_notebook_config.json
|INFO|google.colab serverextension initialized.
jupyter_http_over_ws extension initialized. Listening on /http_over_websocket
|INFO|Serving notebooks from local directory: /
|INFO|The Jupyter Notebook is running at:
|INFO|http://127.0.0.1:8888/
|INFO|Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).

Session terminated, terminating shell...|INFO|interrupted
Serving notebooks from local directory: /
0 active kernels
The Jupyter Notebook is running at:
http://127.0.0.1:8888/
Shutdown this notebook server (y/[n])? y
|CRITICAL|Shutdown confirmed
|INFO|Shutting down 0 kernels
 ...waiting for child to terminate.
 ```

In [ ]:
#!su - engbjapan -c \
!su - engbjapan -c "jupyter notebook \
  --ip=127.0.0.1\
  --port=8888 --FileContentsManager.root_dir=/ --MappingKernelManager.root_dir=/home/engbjapan/ \
  --NotebookApp.allow_origin='https://colab.research.google.com' \
  --NotebookApp.port_retries=0" &

|WARNING|Notebook version 5 is no longer maintained. Please upgrade to version 6 or later.
|INFO|Writing notebook server cookie secret to /home/engbjapan/.local/share/jupyter/runtime/notebook_cookie_secret
|DEBUG|Paths used for configuration of jupyter_notebook_config: 
    	/etc/jupyter/jupyter_notebook_config.json
|DEBUG|Paths used for configuration of jupyter_notebook_config: 
    	/usr/local/etc/jupyter/jupyter_notebook_config.d/panel-client-jupyter.json
    	/usr/local/etc/jupyter/jupyter_notebook_config.json
|DEBUG|Paths used for configuration of jupyter_notebook_config: 
    	/usr/etc/jupyter/jupyter_notebook_config.json
|DEBUG|Paths used for configuration of jupyter_notebook_config: 
    	/home/engbjapan/.local/etc/jupyter/jupyter_notebook_config.json
|DEBUG|Paths used for configuration of jupyter_notebook_config: 
    	/home/engbjapan/.jupyter/jupyter_notebook_config.json
|INFO|google.colab serverextension initialized.
jupyter_http_over_ws extension initialized. Listening on /

# 221119の治験


https://docs.python.org/ja/3.7/library/subprocess.html#subprocess.Popen.communicate

```
proc = subprocess.Popen(...)
try:
    outs, errs = proc.communicate(timeout=15)
except TimeoutExpired:
    proc.kill()
    outs, errs = proc.communicate()
```

In [ ]:
!jupyter console

In [60]:
fd=open("testfile","a")
fd.fileno()

53

In [100]:
!netstat -nap|grep LISTEN

tcp        0      0 127.0.0.1:20624         0.0.0.0:*               LISTEN      95/python3          
tcp        0      0 127.0.0.1:6000          0.0.0.0:*               LISTEN      35/kernel_manager_p 
tcp        0      0 127.0.0.1:41527         0.0.0.0:*               LISTEN      74/python3          
tcp        0      0 127.0.0.1:37113         0.0.0.0:*               LISTEN      95/python3          
tcp        0      0 127.0.0.1:36283         0.0.0.0:*               LISTEN      95/python3          
tcp        0      0 127.0.0.1:40571         0.0.0.0:*               LISTEN      74/python3          
tcp        0      0 127.0.0.1:50747         0.0.0.0:*               LISTEN      74/python3          
tcp        0      0 0.0.0.0:3453            0.0.0.0:*               LISTEN      47/python3          
tcp        0      0 127.0.0.1:45407         0.0.0.0:*               LISTEN      74/python3          
tcp        0      0 127.0.0.1:35167         0.0.0.0:*               LISTEN      74/python3 

だめか・・・

In [97]:
from google.colab import output
output.serve_kernel_port_as_iframe(6000,"/")

<IPython.core.display.Javascript object>

In [79]:
!jupyter kernelspec list

Available kernels:
  bash       /usr/local/share/jupyter/kernels/bash
  ir         /usr/local/share/jupyter/kernels/ir
  python3    /usr/local/share/jupyter/kernels/python3


In [ ]:
!pip install bash_kernel

In [48]:
!python -m bash_kernel.install

Installing IPython kernel spec


In [ ]:
!jupyter console --kernel bash

In [ ]:
import sys
while 1:
  inl=input(">>")
  if inl==":q":
    break
  sys.stdout.write(f"line:{inl}")
  

>>aaaaa
line:aaaaa>>gfggg
line:gfggg

KeyboardInterrupt: ignored

subprocessで、上記起動

stdout/errをopenファイルに。

pidもその時出力か、
戻りのインスタンスで停止する。
```
!su - engbjapan -c "jupyter notebook \
  --ip=127.0.0.1\
  --port=8888 --FileContentsManager.root_dir=/ --MappingKernelManager.root_dir=/home/engbjapan/ \
  --NotebookApp.allow_origin='https://colab.research.google.com' \
  --NotebookApp.port_retries=0" &
```

In [ ]:
class engbshell():
  def __init__(self,user:str="root",input=None,output=None):
    import subprocess ,shlex,os,sys 
    s=self;s.user=user;s.sp=subprocess;s.she=shlex;s.Popen=s.sp.Popen
    self.input=input;self.output=output
  def shell(self,cmds:str):
    s=self
    s.spop = s.Popen(
        s.she.split(cmds.replace('\n',' '))
        
        )


In [ ]:
!ps -Afl > ps.txt
!cat ps.txt

F S UID          PID    PPID  C PRI  NI ADDR SZ WCHAN  STIME TTY          TIME CMD
4 S root           1       0  1  80   0 -   252 sigtim 00:06 ?        00:00:00 /sbin/docker-init -- /datalab/run.sh
4 S root           7       1  1  80   0 - 84443 epoll_ 00:06 ?        00:00:00 /tools/node/bin/node /datalab/web/app.js
4 S root          19       0  1  80   0 -  8975 se_sys 00:06 ?        00:00:00 tail -n +0 -F /root/.config/Google/DriveFS/Logs/dpb.txt /root/.config/Google/DriveFS/Logs/drive_fs.txt
1 S root          33       1  0  80   0 -  9801 wait   00:06 ?        00:00:00 /bin/bash -e /datalab/run.sh
4 S root          35      33  0  80   0 - 179541 futex_ 00:06 ?       00:00:00 /usr/colab/bin/kernel_manager_proxy --listen_port=6000 --target_port=9000
4 Z root          46       7 17  80   0 -     0 -      00:06 ?        00:00:05 [python3] <defunct>
4 S root          47       7  1  80   0 - 40376 epoll_ 00:06 ?        00:00:00 python3 /usr/local/bin/colab-fileshim.py
4 S root          6

In [ ]:
!cat /root/.local/share/jupyter/runtime/kernel-3d4f8a8e-37f9-4ed6-90f7-0ddef8fcf7bf.json

{
  "shell_port": 35167,
  "iopub_port": 58503,
  "stdin_port": 41527,
  "control_port": 50747,
  "hb_port": 45407,
  "ip": "127.0.0.1",
  "key": "",
  "transport": "tcp",
  "signature_scheme": "hmac-sha256",
  "kernel_name": ""
}

# ps
```
!ps -Afl > ps.txt
!cat ps.txt
```
```
F S UID          PID    PPID  C PRI  NI ADDR SZ WCHAN  STIME TTY          TIME CMD
4 S root           1       0  0  80   0 -   252 sigtim 01:50 ?        00:00:00 /sbin/docker-init -- /datalab/run.sh
4 S root           6       1  0  80   0 - 84315 epoll_ 01:50 ?        00:00:00 /tools/node/bin/node /datalab/web/app.js
4 S root          17       0  0  80   0 -  8974 se_sys 01:50 ?        00:00:00 tail -n +0 -F /root/.config/Google/DriveFS/Logs/dpb.txt /root/.config/Google/DriveFS/Logs/drive_fs.txt
1 S root          31       1  0  80   0 -  9801 wait   01:50 ?        00:00:00 /bin/bash -e /datalab/run.sh
4 S root          32      31  0  80   0 - 179953 epoll_ 01:50 ?       00:00:00 /usr/colab/bin/kernel_manager_proxy --listen_port=6000 --target_port=9000
4 Z root          44       6  5  80   0 -     0 -      01:50 ?        00:00:05 [python3] <defunct>
4 S root          45       6  0  80   0 - 40377 epoll_ 01:50 ?        00:00:00 python3 /usr/local/bin/colab-fileshim.py
4 S root          60       6  1  80   0 - 49545 epoll_ 01:50 ?        00:00:01 /usr/bin/python3 /usr/local/bin/jupyter-notebook --ip=172.28.0.2 --port=9000 --FileContentsManager.root_dir=/ --MappingKernelManager.root_dir=/content
4 S root          61       6  0  80   0 - 178533 futex_ 01:50 ?       00:00:00 /usr/local/bin/dap_multiplexer --domain_socket_path=/tmp/debugger_nxw55y8j
4 S root          74      60 13  80   0 - 116165 select 01:52 ?       00:00:01 /usr/bin/python3 -m ipykernel_launcher -f /root/.local/share/jupyter/runtime/kernel-119e19ff-06aa-44f5-b8f6-1e7ed07bbfba.json
1 S root          92       1  0  80   0 - 23283 futex_ 01:52 ?        00:00:00 /usr/bin/python3 /usr/local/lib/python3.7/dist-packages/debugpy/adapter --for-server 36863 --host 127.0.0.1 --port 21946 --server-access-token f1b0b15d8a590be1ec3fe7efcb40e8abef22fe8ae9d75dbbe6bc0c8974df129d
4 R root         109       6  0  80   0 - 68992 -      01:52 ?        00:00:00 node /datalab/web/pyright/pyright-langserver.js --stdio --cancellationReceive=file:baf0d2391abc5a9aefea1986ed6fa989cc6255716e
4 S root         116      74  0  80   0 -  9802 wait   01:52 ?        00:00:00 /bin/bash -c ps -Afl > ps.txt
4 R root         117     116  0  80   0 - 14761 -      01:52 ?        00:00:00 ps -Afl
```

# プロセスの層
1:/sbin/docker-init -- /datalab/run.sh
```
#!/bin/bash -e

# Copyright 2017 Google Inc. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#  http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

if [[ -e /customize_for_vm_type.sh ]]; then
  /customize_for_vm_type.sh
fi

# Start the Colab proxy to the Jupyter kernel manager.
( while true; do
  GCE_METADATA_HOST="${VM_GCE_METADATA_HOST}" \
  /usr/colab/bin/kernel_manager_proxy \
    --listen_port="${KMP_LISTEN_PORT}" \
    --target_port="${KMP_TARGET_PORT}" \
    ${KMP_EXTRA_ARGS} || true
  sleep 1
done & )

# Start fresh to isolate user-initiated actions from VM build & startup events.
for f in /var/log/apt/history.log /var/log/pip.log; do
  mv "$f" "${f}.bak-run.sh" || true  # Ignore missing files.
done

# Warm disk buffers for modules we need for kernel startup. (cf: b/116536906)
python3 -c "import google.colab._kernel"
python3 -c "import matplotlib"
python3 -c "import tensorflow" &

# Start the server to handle /files and /api/contents requests.
/usr/local/bin/colab-fileshim.py &

# Link NVidia tools from a read-only volume mount.
for f in $(ls /opt/bin/.nvidia 2>/dev/null); do
  ln -st /opt/bin ".nvidia/${f}"
done

cd /

# Start the node server.
exec /tools/node/bin/node /datalab/web/app.js
```
↓
→ 6:/tools/node/bin/node /datalab/web/app.js

　　↓
→45:python3 /usr/local/bin/colab-fileshim.py

　　↓
→60:/usr/bin/python3 /usr/local/bin/jupyter-notebook --ip=172.28.0.2 --port=9000 --FileContentsManager.root_dir=/ --MappingKernelManager.root_dir=/content

　　　　↓
→74:/usr/bin/python3 -m ipykernel_launcher -f ***/root/.local/share/jupyter/runtime/kernel-119e19ff-06aa-44f5-b8f6-1e7ed07bbfba.json***

　　　　　　↓
→116:/bin/bash -c ps -Afl > ps.txt

　　　　　　　↓
→117:ps -Afl

　　↓
→61:/usr/local/bin/dap_multiplexer --domain_socket_path=/tmp/debugger_nxw55y8j



　　↓
→109:node /datalab/web/pyright/pyright-langserver.js --stdio --cancellationReceive=file:baf0d2391abc5a9aefea1986ed6fa989cc6255716e

↓→31:/bin/bash -e /datalab/run.sh

　　↓
→32:/usr/colab/bin/kernel_manager_proxy --listen_port=6000 --target_port=9000

↓→:92/usr/bin/python3 /usr/local/lib/python3.7/dist-packages/debugpy/adapter --for-server 36863 --host 127.0.0.1 --port 21946 --server-access-token f1b0b15d8a590be1ec3fe7efcb40e8abef22fe8ae9d75dbbe6bc0c8974df129d


**/root/.local/share/jupyter/runtime/kernel-119e19ff-06aa-44f5-b8f6-1e7ed07bbfba.json**の中身
```
{
  "shell_port": 54273,
  "iopub_port": 50447,
  "stdin_port": 49571,
  "control_port": 51217,
  "hb_port": 52829,
  "ip": "127.0.0.1",
  "key": "",
  "transport": "tcp",
  "signature_scheme": "hmac-sha256",
  "kernel_name": ""
}
```
# 上記portのプロセスは皆、同じで
```
!netstat -nap|egrep '(:54273|:50447|:49571|:51217|:52829)'
```
```
tcp        0      0 127.0.0.1:52829         0.0.0.0:*               LISTEN      74/python3          
tcp        0      0 127.0.0.1:54273         0.0.0.0:*               LISTEN      74/python3          
tcp        0      0 127.0.0.1:49571         0.0.0.0:*               LISTEN      74/python3          
tcp        0      0 127.0.0.1:50447         0.0.0.0:*               LISTEN      74/python3          
tcp        0      0 127.0.0.1:51217         0.0.0.0:*               LISTEN      74/python3          
tcp        0      0 127.0.0.1:50447         127.0.0.1:39686         ESTABLISHED 74/python3          
tcp        0      0 127.0.0.1:39686         127.0.0.1:50447         ESTABLISHED 60/python3          
tcp        0    460 127.0.0.1:49116         127.0.0.1:54273         ESTABLISHED 60/python3          
tcp        0      0 127.0.0.1:51217         127.0.0.1:33162         ESTABLISHED 74/python3          
tcp        0      0 127.0.0.1:50447         127.0.0.1:39674         ESTABLISHED 74/python3          
tcp        0      0 127.0.0.1:56674         127.0.0.1:49571         ESTABLISHED 60/python3          
tcp        0      0 127.0.0.1:39674         127.0.0.1:50447         ESTABLISHED 60/python3          
tcp        0      0 127.0.0.1:49571         127.0.0.1:56674         ESTABLISHED 74/python3          
tcp        0      0 127.0.0.1:54273         127.0.0.1:49116         ESTABLISHED 74/python3          
tcp        0      0 127.0.0.1:33162         127.0.0.1:51217         ESTABLISHED 60/python3 
```


ipykernel_launcher

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py

# また、以下のPORTプロセス層もどうなんだ！！

/usr/colab/bin/kernel_manager_proxy --listen_port=6000 --target_port=9000

で上記target_port=9000はjupyter notebook起動時のPORT

これは、127系のポートで動くのを何等かProxyをlocalからの要求転送してると思われる

```
tcp        0      0 127.0.0.1:6000          0.0.0.0:*               LISTEN      33/kernel_manager_p 
tcp        0      0 172.28.0.2:9000         0.0.0.0:*               LISTEN      60/python3          
tcp        0      0 172.28.0.2:9000         172.28.0.3:46126        ESTABLISHED 60/python3          
tcp        0      0 172.28.0.2:9000         172.28.0.3:35172        ESTABLISHED 60/python3          
tcp        0      0 172.28.0.2:37256        172.28.0.3:6000         ESTABLISHED 6/node      
```
node ⇒172.28.0.3:6000
  
  ⇒127.0.0.1:6000
  
  　　⇒/usr/colab/bin/kernel_manager_proxy --listen_port=6000 --target_port=9000

  　　　⇒172.28.0.2:9000

  　　　　⇒/usr/bin/python3 /usr/local/bin/jupyter-notebook --ip=172.28.0.2 --port=9000 --FileContentsManager.root_dir=/ --MappingKernelManager.root_dir=/content

---
172.28.0.3って・・・

```
eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 172.28.0.2  netmask 255.255.0.0  broadcast 172.28.255.255
        ether 02:42:ac:1c:00:02  txqueuelen 0  (Ethernet)
        RX packets 12992  bytes 4230043 (4.2 MB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 11402  bytes 2949367 (2.9 MB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 23150  bytes 5684326 (5.6 MB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 23150  bytes 5684326 (5.6 MB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

# nodeで起動の /datalab/web/app.js　の中身

```
"use strict";
/*
 * Copyright 2015 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("fs");
var path = require("path");
var logging = require("./logging");
var server = require("./server");
/**
 * Loads the configuration settings for the application to use.
 * On first run, this generates any dynamic settings and merges them into the
 * settings result.
 * @returns the settings object for the application to use.
 */
function loadAppSettings() {
    var settingsPath = path.join(__dirname, 'config', 'settings.json');
    if (!fs.existsSync(settingsPath)) {
        var msg = "App settings file \"".concat(settingsPath, "\" not found.");
        console.error(msg);
        throw new Error(msg);
    }
    try {
        var settings = JSON.parse(fs.readFileSync(settingsPath, 'utf8') || '{}');
        var settingsOverrides = process.env['DATALAB_SETTINGS_OVERRIDES'];
        if (settingsOverrides) {
            // Allow overriding individual settings via JSON provided as an environment variable.
            var overrides = JSON.parse(settingsOverrides);
            Object.assign(settings, overrides);
        }
        return settings;
    }
    catch (e) {
        console.error(e);
        throw new Error("Error parsing settings overrides: ".concat(e));
    }
}
/**
 * Load the configuration settings, and then start the server, which
 * runs indefinitely, listening to and processing incoming HTTP requests.
 */
var appSettings = loadAppSettings();
if (appSettings != null) {
    logging.initializeLoggers(appSettings);
    server.run(appSettings);
}
/**
 * Handle shutdown of this process, to also stop the server, which will in turn stop the
 * associated Jupyter server process.
 */
function exit() {
    logging.getLogger().info('app: exit');
    server.stop();
    logging.getLogger().info('app: exit: stopped');
    process.exit(0);
}
/**
 * Handle uncaught exceptions to log them.
 */
function errorHandler(e) {
    console.error(e.stack);
    logging.getLogger().error(e, 'Unhandled exception');
    process.exit(1);
}
process.on('uncaughtException', errorHandler);
process.on('exit', exit);
process.on('SIGINT', exit);
process.on('SIGTERM', exit);
//# sourceMappingURL=data:application/json;base64,eyJ2....ZXhpdCk7XG4iXX0=
```

In [ ]:
!ls -la /

# 以下で　ipython profileディレクトリが見れるそうだ

```
!ls -la `ipython profile locate default`
```
```
total 132
drwxr-xr-x 7 root root  4096 Nov 15 14:47 .
drwxr-xr-x 5 root root  4096 Nov 15 14:46 ..
drwxr-xr-x 2 root root  4096 Nov 15 14:47 db
-rw-r--r-- 1 root root 28672 Nov 15 14:47 history.sqlite
-rw-r--r-- 1 root root 38474 Nov 15 14:46 ipython_config.py
-rw-r--r-- 1 root root 33065 Nov 15 14:46 ipython_kernel_config.py
drwxr-xr-x 2 root root  4096 Nov 15 14:46 log
drwx------ 2 root root  4096 Nov 15 14:46 pid
drwx------ 2 root root  4096 Nov 15 14:46 security
drwxr-xr-x 2 root root  4096 Nov 15 14:46 startup
```
##他にもipythonの実行で稼働が見れるようだ、試してみよう！！


In [ ]:
!/usr/bin/python3 /usr/local/bin/jupyter-notebook --ip=172.28.0.2 --port=8888 --FileContentsManager.root_dir=/ --MappingKernelManager.root_dir=/home/engbjapan/

# さて、拡張使う前にyupyterの起動は以下で一応したが。
Tokenが表示されず、バックグラウンドで動かないっからこれも解決するよ！

https://ipython.org/ipython-doc/3/config/options/notebook.html
は？tokenの起動時指定が見当たらない

後は、ダイレクトに
jupyter notebook --create-conmfig
で作ったconfigにTokenを適当に設定する手：

```
!su - engbjapan -c "/usr/bin/python3 /usr/local/bin/jupyter-notebook \
--ip=127.0.0.1 --port=8888 --FileContentsManager.root_dir=/ \
--MappingKernelManager.root_dir=/home/engbjapan/ --NotebookApp.allow_origin='https://colab.research.google.com'"
```
--ip=127.0.0.1 指定がSocketBiond出来たから必須

```
|WARNING|Notebook version 5 is no longer maintained. Please upgrade to version 6 or later.
***Debug***
        sockaddr('127.0.0.1', 8888)
|DEBUG|Paths used for configuration of jupyter_notebook_config: 
    	/etc/jupyter/jupyter_notebook_config.json
|DEBUG|Paths used for configuration of jupyter_notebook_config: 
    	/usr/local/etc/jupyter/jupyter_notebook_config.d/panel-client-jupyter.json
    	/usr/local/etc/jupyter/jupyter_notebook_config.json
|DEBUG|Paths used for configuration of jupyter_notebook_config: 
    	/usr/etc/jupyter/jupyter_notebook_config.json
|DEBUG|Paths used for configuration of jupyter_notebook_config: 
    	/home/engbjapan/.local/etc/jupyter/jupyter_notebook_config.json
|DEBUG|Paths used for configuration of jupyter_notebook_config: 
    	/home/engbjapan/.jupyter/jupyter_notebook_config.json
|INFO|google.colab serverextension initialized.
jupyter_http_over_ws extension initialized. Listening on /http_over_websocket
|INFO|Serving notebooks from local directory: /
|INFO|The Jupyter Notebook is running at:
|INFO|**http://127.0.0.1:8888/**
|INFO|Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).

Session terminated, terminating shell...|INFO|interrupted
Serving notebooks from local directory: /
0 active kernels
The Jupyter Notebook is running at:
**http://127.0.0.1:8888/**
Shutdown this notebook server (y/[n])? No answer for 5s: resuming operation...
y
```

# jupyter connfig 見る
```
!find / -name "*jupyter_notebook_config.py" 2>/dev/nul|while read f;do printf "file:${f}\n=====\n";cat $f|egrep '^[^\#]+.*$';done
```
```
file:/root/.jupyter/jupyter_notebook_config.py
=====
file:/etc/jupyter/jupyter_notebook_config.py
=====
"""Jupyter notebook server configuration for Colab runtimes."""
import logging
c = get_config()  # pylint: disable=undefined-variable
c.JupyterApp.answer_yes = True
c.NotebookApp.open_browser = False
c.NotebookApp.allow_root = True
c.Application.log_level = logging.DEBUG
c.NotebookApp.log_format = '|%(levelname)s|%(message)s'
c.NotebookApp.port_retries = 0
c.NotebookApp.login_handler_class = (
    'google.colab._login_handler.ColabLoginHandler')
c.MultiKernelManager.default_kernel_name = 'python3'
c.NotebookApp.allow_origin = '*'
c.NotebookApp.allow_remote_access = True
c.NotebookApp.disable_check_xsrf = True
c.NotebookApp.token = ''
c.Session.key = b''
c.Session.keyfile = ''
```

su - {ユーザ} -c 'jupyter notebook --create-config'

して

「/etc/jupyter/jupyter_notebook_config.py」のCopyを/home/{ユーザ}/.jupyter/以下に

c.NotebookApp.token = 'XXXXXXXXX'
にtoken指定して
su - engbjapan -c 'jupyter notebook ...'
を実行では？？
